libkbfs/folder_branch_ops.go

```diff
@@ -3179,101 +3207,24 @@ func (fbo *folderBranchOps) renameLocked(
-	for i = 1; i < len(oldParentPath.path) && i < len(newParentPath.path); i++ {
-		if oldParentPath.path[i].ID != newParentPath.path[i].ID {
-			found = true
-			i--
-			break
-		}
-	}
-	if !found {
-		// if we couldn't find one, then the common ancestor is the
-		// last node in the shorter path
-		if len(oldParentPath.path) < len(newParentPath.path) {
-			i = len(oldParentPath.path) - 1
-		} else {
-			i = len(newParentPath.path) - 1
-		}
-	}
-	commonAncestor := oldParentPath.path[i].BlockPointer
-	oldIsCommon := oldParentPath.tailPointer() == commonAncestor
-	newIsCommon := newParentPath.tailPointer() == commonAncestor
-
-	newOldPath := path{FolderBranch: oldParentPath.FolderBranch}
-	var oldBps *blockPutState
-	if oldIsCommon {
-		if newIsCommon {
-			// if old and new are both the common ancestor, there is
-			// nothing to do (syncBlock will take care of everything)
-		} else {
-			// If the old one is common and the new one is
-			// not, then the last
-			// syncBlockAndCheckEmbedLocked call will need
-			// to access the old one.
-			lbc[oldParentPath.tailPointer()] = oldPBlock
-		}
-	} else {
-		if newIsCommon {
-			// If the new one is common, then the first
-			// syncBlockAndCheckEmbedLocked call will need to access
-			// it.
-			lbc[newParentPath.tailPointer()] = newPBlock
-		}
-
-		// The old one is not the common ancestor, so we need to sync it.
-		// TODO: optimize by pushing blocks from both paths in parallel
-		newOldPath, _, oldBps, err = fbo.syncBlockAndCheckEmbedLocked(
-			ctx, lState, md, oldPBlock, *oldParentPath.parentPath(),
-			oldParentPath.tailName(), Dir, true, true, commonAncestor, lbc)
-		if err != nil {
-			return err
-		}
-	}
-
-	newNewPath, _, newBps, err := fbo.syncBlockAndCheckEmbedLocked(
-		ctx, lState, md, newPBlock, *newParentPath.parentPath(),
-		newParentPath.tailName(), Dir, true, true, zeroPtr, lbc)
-	if err != nil {
-		return err
-	}
-
-	// newOldPath is really just a prefix now.  A copy is necessary as an
-	// append could cause the new path to contain nodes from the old path.
-	newOldPath.path = append(make([]pathNode, i+1, i+1), newOldPath.path...)
-	copy(newOldPath.path[:i+1], newNewPath.path[:i+1])
-
-	// merge and finalize the blockPutStates
-	if oldBps != nil {
-		newBps.mergeOtherBps(oldBps)
-	}
-
-	defer func() {
 		if err != nil {
-			fbo.fbm.cleanUpBlockState(
-				md.ReadOnly(), newBps, blockDeleteOnMDFail)
+			for _, n := range addedNodes {
+				fbo.status.rmDirtyNode(n)
+			}
+			fbo.dirOps = fbo.dirOps[:len(fbo.dirOps)-1]
+			dirCacheUndoFn(lState)
 		}
 	}()
 
-	_, err = doBlockPuts(ctx, fbo.config.BlockServer(), fbo.config.BlockCache(),
-		fbo.config.Reporter(), fbo.log, md.TlfID(),
-		md.GetTlfHandle().GetCanonicalName(), *newBps)
+	// It's safe to notify before we've synced, since it is only
+	// sending invalidation notifications.  At worst the upper layer
+	// will just have to refresh its cache needlessly.
+	err = fbo.notifyOneOp(ctx, lState, ro, md.ReadOnly(), false)
```

Comment:
The above defer through to this `notifyOneOp` is repeated at least 3-4 times (except for the specific undo function that's called). Is there a way we could factor it out a bit?